Alerting: Add label and state filters to alert instance components

JacobsonMT opened the pull request:
**What this PR does / why we need it**:

Adds label and state filtering to all components that use AlertInstancesTable, this includes:

- Alert List panel (this panel was standardized to use AlertInstancesTable)
- Alerting / View rule
- Alerting / Alert rules
- Panel-specific alert tab

Also modifies RadioButton with optional onClick to allow for RadioButtonGroup toggling on second click

**Screenshots / gifs**:

![Peek 2021-11-30 18-01](https://user-images.githubusercontent.com/8484471/144144019-ac633d0f-c23e-4860-b620-6aef2238e87a.gif)

![Peek 2021-11-30 18-13](https://user-images.githubusercontent.com/8484471/144144024-18cd5864-b983-48cc-8941-41ef0b4802da.gif)

![Screenshot from 2021-11-30 18-28-49](https://user-images.githubusercontent.com/8484471/144145287-b7522149-b0ab-46ad-a78a-3587f0548662.png)


**Which issue(s) this PR fixes**:

Fixes #36048
Relates to #35720

JacobsonMT committed:
Alerting: Add label and state filters to alert instance components

Adds label and state filtering to all components that use
AlertInstancesTable, this includes:
    - Alert List panel (this panel was standardized to use AlertInstancesTable)
    - Alerting / View rule
    - Alerting / Alert rules
    - Panel-specific alert tab

Modifies RadioButton with optional onClick to allow for RadioButtonGroup toggling

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=42550) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=42550) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=42550) it.</sub>

JohnnyQQQQ commented:
@JacobsonMT thanks for your first contribution to Grafana and welcome to the community 🚀   I did approve the drone build stage and added some folks that will review this frontend change.

gillesdemey reviewed (changes requested):
Since this PR is quite large I'm releasing my first batch of comments; some of these are not blockers and can be improved upon in a follow-up PR.

A few blocking issues though, those are not preceded by `nit:` 💡

JacobsonMT commented on public/app/features/alerting/unified/components/rules/RuleDetailsMatchingInstances.tsx:
```diff
@@ -11,13 +18,126 @@ type Props = {
+    [promRule, alertState, queryString]
+  );
+
+  if (!alerts.length) {
```
Good catch, thanks! That's what I get for not retesting after fixing pre-commit errors :)

JacobsonMT commented on public/app/features/alerting/unified/components/rules/RuleDetailsMatchingInstances.tsx:
```diff
@@ -11,13 +18,126 @@ type Props = {
 export function RuleDetailsMatchingInstances(props: Props): JSX.Element | null {
   const { promRule } = props;
 
-  if (!isAlertingRule(promRule) || !promRule.alerts?.length) {
```
So, for recording rules we should not display the table/filters at all (return null)?

gillesdemey replied:
Yep exactly, that is also why the `RuleDetailsMatchingInstances` was typed as `JSX.Element | null` :v:

gillesdemey commented on public/app/features/alerting/unified/utils/alertmanager.ts:
```diff
@@ -136,7 +136,7 @@ export function parseMatcher(matcher: string): Matcher {
 
 export function parseMatchers(matcherQueryString: string): Matcher[] {
-  const matcherRegExp = /\b(\w+)(=~|!=|!~|=(?="?\w))"?([^"\n,]*)"?/g;
+  const matcherRegExp = /\b([\w.-]+)(=~|!=|!~|=(?="?\w))"?([^"\n,]*)"?/g;
```
Awesome! TIL that the `.` is literal in a character class!

As far as suggestions go I can think of the underscore (`_`) character if we aren't including that one yet, that should do it for now :)

JacobsonMT commented on packages/grafana-ui/src/components/Forms/RadioButtonGroup/RadioButtonGroup.tsx:
```diff
@@ -23,6 +24,7 @@ export function RadioButtonGroup<T>({
   options,
   value,
   onChange,
+  onClick,
```
Implying adding this new "unselect an item by re-clicking it" as the default for RadioButtonGroup? I think it's a good idea, I mostly did it this way as I was a bit wary of modifying the default functionality of something used in so many places of the UI.

I'll give it a shot and see if I notice any strangeness as a result.

JacobsonMT committed:
Alerting: Extract alert instance filters into components

RuleDetailsMatchingInstances filters extracted into:
- MatcherFilter (existing)
- AlertInstanceStateFilter (new)

JacobsonMT committed:
Chore: Fix ts-ignore for alert instance score sorting

JacobsonMT committed:
Alerting: Ensure recording rules return null RuleDetailsMatchingInstances

JacobsonMT commented:
Addressed all comments except for "onUnselect" as I'd like to test it a bit more locally

gillesdemey reviewed (approved):
Amazing first-time contribution, LGTM! :shipit:
